src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs

```diff
@@ -889,6 +934,43 @@ private PackageLicenseInfo GeneratePackageLicenseInfo(IPackageSearchMetadata met
             return results;
         }
 
+        private static void CreateAndSavePackageSourceMappings(string sourceName, string[] newPackageIdsToSourceMap, PackageSourceMappingProvider mappingProvider)
+        {
+            if (string.IsNullOrWhiteSpace(sourceName) || newPackageIdsToSourceMap is null || newPackageIdsToSourceMap.Length == 0)
+            {
+                return;
+            }
+
+            IEnumerable<PackagePatternItem> newPackagePatternItems = newPackageIdsToSourceMap.Select(packageId => new PackagePatternItem(packageId));
+
+            IReadOnlyList<PackageSourceMappingSourceItem> existingPackageSourceMappingItems = mappingProvider.GetPackageSourceMappingItems();
+            List<PackageSourceMappingSourceItem> newAndExistingPackageSourceMappingItems = new(existingPackageSourceMappingItems);
+
+            PackageSourceMappingSourceItem existingPackageSourceMappingItemForSource =
+                        existingPackageSourceMappingItems
+                        .Where(mappingItem => mappingItem.Key == sourceName)
+                        .FirstOrDefault();
+
+            // Source is being mapped for the first time.
+            if (existingPackageSourceMappingItemForSource is null)
+            {
+                existingPackageSourceMappingItemForSource = new(sourceName, newPackagePatternItems);
+                newAndExistingPackageSourceMappingItems.Add(existingPackageSourceMappingItemForSource);
+            }
+            else // Source already had an existing mapping.
```

Comment:
A few lines above where this method is called, a variable named `packageIdsNeedingNewSourceMappings` is defined, with a comment saying that it's already calculated packages that don't already have source mappings.

I haven't read this code carefully enough to figure out what the answer is, but are "package IDs that don't already have mappings" being calculated twice? Or is one of the two comments no longer correct?

Reply:
Yes, I think I was retrieving existing mappings again in the save method, and technically those shouldn't be modified during the preview restore, so now I am passing them down into the save.